File: fBReader/src/main/java/org/geometerplus/fbreader/bookmodel/BookModel.java
/*
 * Copyright (C) 2007-2015 FBReader.ORG Limited <contact@fbreader.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

package org.geometerplus.fbreader.bookmodel;

import org.geometerplus.fbreader.book.Book;
import org.geometerplus.fbreader.formats.BookReadingException;
import org.geometerplus.fbreader.formats.BuiltinFormatPlugin;
import org.geometerplus.fbreader.formats.FormatPlugin;
import org.geometerplus.zlibrary.core.fonts.FileInfo;
import org.geometerplus.zlibrary.core.fonts.FontEntry;
import org.geometerplus.zlibrary.core.fonts.FontManager;
import org.geometerplus.zlibrary.core.image.ZLImage;
import org.geometerplus.zlibrary.text.model.CachedCharStorage;
import org.geometerplus.zlibrary.text.model.ZLTextModel;
import org.geometerplus.zlibrary.text.model.ZLTextPlainModel;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import timber.log.Timber;

public final class BookModel {

	/**
	 * 1. 创建book渲染数据model第一步
	 */
	public static BookModel createModel(Book book, FormatPlugin plugin) throws BookReadingException {
		Timber.v("图书解析流程, 开始解析图书, 创建图书model, %s",  plugin.getClass().getSimpleName());
		// 只有FB2NativePlugin和OEBNativePlugin才会执行cpp层解析
		if (!(plugin instanceof BuiltinFormatPlugin)) {
			throw new BookReadingException(
					"unknownPluginType", null, new String[]{String.valueOf(plugin)}
			);
		}

		// 保存图书基本信息: title, path, encoding,
		final BookModel model = new BookModel(book);
		// 图书解析入口, 调用cpp层开始图书解析
		((BuiltinFormatPlugin) plugin).readModel(model);
		return model;
	}

	public final Book Book;
	public final TOCTree TOCTree = new TOCTree();
	public final FontManager FontManager = new FontManager();

	protected CachedCharStorage myInternalHyperlinks;
	protected final HashMap<String,ZLImage> myImageMap = new HashMap<>();
	protected ZLTextModel myBookTextModel;
	protected final HashMap<String,ZLTextModel> myFootnotes = new HashMap<>();

	public static final class Label {
		public final String ModelId;
		public final int ParagraphIndex;

		public Label(String modelId, int paragraphIndex) {
			ModelId = modelId;
			ParagraphIndex = paragraphIndex;
		}
	}

	protected BookModel(Book book) {
		Book = book;
	}

	public interface LabelResolver {
		List<String> getCandidates(String id);
	}

	private LabelResolver myResolver;

	public void setLabelResolver(LabelResolver resolver) {
		myResolver = resolver;
	}

	/**
	 * 获得footNote/超链接数据
	 */
	public Label getLabel(String id) {
		Timber.v("超链接, id = %s", id);
		Label label = getLabelInternal(id);
		if (label == null && myResolver != null) {
			for (String candidate : myResolver.getCandidates(id)) {
				label = getLabelInternal(candidate);
				if (label != null) {
					break;
				}
			}
		}
		return label;
	}

	public ZLTextModel getTextModel() {
		return myBookTextModel;
	}

	public ZLTextModel getFootnoteModel(String id) {
		return myFootnotes.get(id);
	}

	public void addImage(String id, ZLImage image) {
		myImageMap.put(id, image);
	}

	private Label getLabelInternal(String id) {
		final int len = id.length();
		final int size = myInternalHyperlinks.size();

		for (int i = 0; i < size; ++i) {
			final char[] block = myInternalHyperlinks.block(i);
			for (int offset = 0; offset < block.length; ) {
				final int labelLength = block[offset++];
				if (labelLength == 0) {
					break;
				}
				final int idLength = (int)block[offset + labelLength];
				if (labelLength != len || !id.equals(new String(block, offset, labelLength))) {
					offset += labelLength + idLength + 3;
					continue;
				}
				offset += labelLength + 1;
				final String modelId = (idLength > 0) ? new String(block, offset, idLength) : null;
				offset += idLength;
				final int paragraphNumber = (int)block[offset] + (((int)block[offset + 1]) << 16);
				return new Label(modelId, paragraphNumber);
			}
		}
		return null;
	}

	/********************************** 以下为cpp调用java的方法 *******************************/
	public void registerFontFamilyList(String[] families) {
		FontManager.index(Arrays.asList(families));
	}

	public void registerFontEntry(String family, FontEntry entry) {
		FontManager.Entries.put(family, entry);
	}

	public void registerFontEntry(String family, FileInfo normal, FileInfo bold, FileInfo italic, FileInfo boldItalic) {
		registerFontEntry(family, new FontEntry(family, normal, bold, italic, boldItalic));
	}

	public ZLTextModel createTextModel(
			String id, String language, int paragraphsNumber,
			int[] entryIndices, int[] entryOffsets,
			int[] paragraphLenghts, int[] textSizes, byte[] paragraphKinds,
			String directoryName, String fileExtension, int blocksNumber
	) {
		return new ZLTextPlainModel(
				id, language, paragraphsNumber,
				entryIndices, entryOffsets,
				paragraphLenghts, textSizes, paragraphKinds,
				directoryName, fileExtension, blocksNumber, myImageMap, FontManager
		);
	}

	public void setBookTextModel(ZLTextModel model) {
		Timber.v("cpp解析打印, cpp调用java方法");
		myBookTextModel = model;
	}

	public void setFootnoteModel(ZLTextModel model) {
		myFootnotes.put(model.getId(), model);
	}

	public void initInternalHyperlinks(String directoryName, String fileExtension, int blocksNumber) {
		myInternalHyperlinks = new CachedCharStorage(directoryName, fileExtension, blocksNumber);
	}

	private TOCTree myCurrentTree = TOCTree;

	public void addTOCItem(String text, int reference) {
		myCurrentTree = new TOCTree(myCurrentTree);
		myCurrentTree.setText(text);
		myCurrentTree.setReference(myBookTextModel, reference);
	}

	public void leaveTOCItem() {
		myCurrentTree = myCurrentTree.Parent;
		if (myCurrentTree == null) {
			myCurrentTree = TOCTree;
		}
	}
	/********************************** 以上为cpp调用java的方法 *******************************/

	@Override
	public String toString() {
		return "BookModel{" +
				"\nBook=" + Book +
				", \nTOCTree=" + TOCTree +
				", \nFontManager=" + FontManager +
				", \nmyInternalHyperlinks=" + myInternalHyperlinks +
				", \nmyImageMap=" + myImageMap +
				", \nmyBookTextModel=" + myBookTextModel +
				", \nmyFootnotes=" + myFootnotes +
				", \nmyResolver=" + myResolver +
				", \nmyCurrentTree=" + myCurrentTree +
				'}';
	}
}

File: fBReader/src/main/java/org/geometerplus/fbreader/bookmodel/BookModel.kt
package org.geometerplus.fbreader.bookmodel

import org.geometerplus.fbreader.book.Book
import org.geometerplus.fbreader.formats.BookReadingException
import org.geometerplus.fbreader.formats.BuiltinFormatPlugin
import org.geometerplus.fbreader.formats.FormatPlugin
import org.geometerplus.zlibrary.core.fonts.FileInfo
import org.geometerplus.zlibrary.core.fonts.FontEntry
import org.geometerplus.zlibrary.core.fonts.FontManager
import org.geometerplus.zlibrary.core.image.ZLImage
import org.geometerplus.zlibrary.text.model.CachedCharStorage
import org.geometerplus.zlibrary.text.model.ZLTextModel
import org.geometerplus.zlibrary.text.model.ZLTextPlainModel
import timber.log.Timber
import java.util.*

/**
 * @Package org.geometerplus.fbreader.bookmodel
 * @FileName BookModel
 * @Date 5/28/23, 2:19 PM
 * @Author Created by fengchengding
 * @Description FBReaderJ_AndroidStudio
 */
class BookModel(
    @get:JvmName("book") val book: Book?,
    @get:JvmName("tocTree") val tocTree: TOCTree = TOCTree(),
) {

    private val fontManager = FontManager()

    private var internalHyperlinks: CachedCharStorage? = null
    private val imageMap = mutableMapOf<String, ZLImage>()
    private var bookTextModel: ZLTextModel? = null
    private val footnotes = mutableMapOf<String, ZLTextModel>()
    private var labelResolver: LabelResolver? = null

    interface LabelResolver {
        fun getCandidates(id: String?): List<String>
    }

    fun setLabelResolver(resolver: LabelResolver) {
        labelResolver = resolver
    }

    /**
     * 获得footNote/超链接数据
     */
    fun getLabel(id: String): Label? {
        Timber.v("超链接, id = %s", id)
        var label = getLabelInternal(id)
        if (label == null && labelResolver != null) {
            for (candidate in labelResolver!!.getCandidates(id)) {
                label = getLabelInternal(candidate)
                if (label != null) {
                    break
                }
            }
        }
        return label
    }

    fun getTextModel(): ZLTextModel? {
        return bookTextModel
    }

    fun getFootnoteModel(id: String): ZLTextModel? {
        return footnotes[id]
    }

    fun addImage(id: String, image: ZLImage) {
        imageMap[id] = image
    }

    private fun getLabelInternal(id: String): Label? {
        val len = id.length
        val size = internalHyperlinks!!.size()
        for (i in 0 until size) {
            val block = internalHyperlinks!!.block(i)
            var offset = 0
            while (offset < block.size) {
                val labelLength = block[offset++].code
                if (labelLength == 0) {
                    break
                }
                val idLength = block[offset + labelLength].code
                if (labelLength != len || id != String(block, offset, labelLength)) {
                    offset += labelLength + idLength + 3
                    continue
                }
                offset += labelLength + 1
                val modelId = if (idLength > 0) String(block, offset, idLength) else null
                offset += idLength
                val paragraphNumber = block[offset].code + (block[offset + 1].code shl 16)
                return Label(modelId, paragraphNumber)
            }
        }
        return null
    }

    /********************************** 以下为cpp调用java的方法 *******************************/
    fun registerFontFamilyList(families: Array<String>) {
        fontManager.index(families.toList())
    }

    fun registerFontEntry(family: String, entry: FontEntry) {
        fontManager.Entries[family] = entry
    }

    fun registerFontEntry(
        family: String,
        normal: FileInfo,
        bold: FileInfo,
        italic: FileInfo,
        boldItalic: FileInfo
    ) {
        registerFontEntry(family, FontEntry(family, normal, bold, italic, boldItalic))
    }

    fun createTextModel(
        id: String?,
        language: String,
        paragraphsNumber: Int,
        entryIndices: IntArray,
        entryOffsets: IntArray,
        paragraphLenghts: IntArray,
        textSizes: IntArray,
        paragraphKinds: ByteArray,
        directoryName: String,
        fileExtension: String,
        blocksNumber: Int
    ): ZLTextModel =
        ZLTextPlainModel(
            id,
            language,
            paragraphsNumber,
            entryIndices,
            entryOffsets,
            paragraphLenghts,
            textSizes,
            paragraphKinds,
            directoryName,
            fileExtension,
            blocksNumber,
            imageMap,
            fontManager
        )

    fun setBookTextModel(model: ZLTextModel) {
        bookTextModel = model
    }

    fun setFootnoteModel(model: ZLTextModel) {
        footnotes[model.id] = model
    }

    fun initInternalHyperlinks(directoryName: String, fileExtension: String, blocksNumber: Int) {
        internalHyperlinks = CachedCharStorage(directoryName, fileExtension, blocksNumber)
    }

    private var myCurrentTree = tocTree

    fun addTOCItem(text: String, reference: Int) {
        myCurrentTree = TOCTree(myCurrentTree)
        myCurrentTree.text = text
        myCurrentTree.setReference(bookTextModel, reference)
    }

    fun leaveTOCItem() {
        myCurrentTree = myCurrentTree.Parent
        if (myCurrentTree == null) {
            myCurrentTree = tocTree
        }
    }

    /********************************** 以上为cpp调用java的方法 *******************************/

    class Label(
        @get:JvmName("modelId") val modelId: String?,
        @get:JvmName("paragraphIndex") val paragraphIndex: Int
    )

    companion object {

        @JvmStatic
        @Throws(BookReadingException::class)
        fun createModel(book: Book, plugin: FormatPlugin): BookModel {
            Timber.v("图书解析流程, 开始解析图书, 创建图书model, %s", plugin.javaClass.simpleName)
            // 只有FB2NativePlugin和OEBNativePlugin才会执行cpp层解析
            if (plugin !is BuiltinFormatPlugin) {
                throw BookReadingException("unknownPluginType", null, arrayOf("$plugin"))
            }

            // 保存图书基本信息: title, path, encoding,
            val model = BookModel(book)
            // 图书解析入口, 调用cpp层开始图书解析
            plugin.readModel(model)
            return model
        }
    }
}